core/dbt/contracts/graph/metrics.py

```diff
@@ -17,76 +24,74 @@ class ResolvedMetricReference(MetricReference):
     for working with metrics (ie. __str__ and templating functions)
     """
 
-    def __init__(self, node, manifest, Relation):
+    def __init__(self, node: Metric, manifest: Manifest) -> None:
         super().__init__(node.name, node.package_name)
         self.node = node
         self.manifest = manifest
-        self.Relation = Relation
 
-    def __getattr__(self, key):
+    def __getattr__(self, key) -> Any:
         return getattr(self.node, key)
 
-    def __str__(self):
+    def __str__(self) -> str:
         return f"{self.node.name}"
 
     @classmethod
-    def parent_metrics(cls, metric_node, manifest):
+    def parent_metrics(cls, metric_node: Metric, manifest: Manifest) -> Iterator[Metric]:
+        """For a given metric, yeilds all upstream metrics."""
         yield metric_node
 
         for parent_unique_id in metric_node.depends_on.nodes:
-            node = manifest.metrics.get(parent_unique_id)
-            if node and node.resource_type == NodeType.Metric:
-                yield from cls.parent_metrics(node, manifest)
+            metric = manifest.metrics.get(parent_unique_id)
+            if metric is not None:
```

Comment:
is there a valid scenario where it'd be valid for a unique_id to exist in `metric_node.depends_on.nodes` but not be in `manifest.metric`? If not, we could handle this less gracefully by getting metric from `metric = manifest.expect(parent_unique_id)` to ensure an issue is not silently handled.

Reply:
`depends_on` contains metrics and things that aren't metrics (like semantic models). So if we were to use `manifest.expect(...)` we'd have to bring back checking if the returned node is a metric.

Reply:
I see! but once we've got what we know is a metric -- should we expect it to ever be `None`? If not, we could assert/raise an internal error here to make future investigation easier

Reply:
Oooo expect does a look up against all the manifest.x objects. I though this would be a bad runtime cost, but these are dictionaries not lists, so doing an `in` lookup wouldn't be too costly. I'll switch to that as you suggested. That function handles the `None` case, and thus all we need to do on return is check if it's a metric node.